addons/spreadsheet/static/src/chart/plugins/odoo_chart_feature_plugin.js

```diff
@@ -0,0 +1,101 @@
+                        if (fieldMatching.chain === fieldName && bestGranularity !== granularity) {
+                            this.dispatch("UPDATE_CHART_GRANULARITY", {
+                                chartId,
+                                granularity: bestGranularity,
+                            });
+                        }
+                    }
+                }
+                break;
+            }
+            case "UPDATE_CHART_GRANULARITY": {
+                const definition = this.getters.getChartDefinition(cmd.chartId);
+                const { fieldName } = this.getters.getChartGranularity(cmd.chartId);
+                this.dispatch("UPDATE_CHART", {
+                    figureId: cmd.chartId,
+                    definition: {
+                        ...definition,
+                        // I don't know why it's in both searchParams and metaData.
+                        // I think only metaData actually matters.
+                        searchParams: {
+                            ...definition.searchParams,
+                            groupBy: [
+                                `${fieldName}:${cmd.granularity}`,
+                                ...definition.searchParams.groupBy.slice(1),
+                            ],
+                        },
+                        metaData: {
+                            ...definition.metaData,
+                            groupBy: [
+                                `${fieldName}:${cmd.granularity}`,
+                                ...definition.metaData.groupBy.slice(1),
+                            ],
+                        },
+                    },
+                });
+            }
+        }
+    }
+
+    getAvailableChartGranularities(chartId) {
+        const definition = this.getters.getChartDefinition(chartId);
+        if (!definition.type.startsWith("odoo_")) {
+            return [];
+        }
+        const { granularity } = this.getters.getChartGranularity(chartId);
+        if (!granularity) {
+            return [];
+        }
+        const filterId = this._getChartHorizontalAxisFilter(chartId);
+        const currentFilterValue = filterId
+            ? this.getters.getGlobalFilterValue(filterId)
+            : undefined;
+        const all = [
+            { value: "day", label: _t("Days") },
+            { value: "week", label: _t("Weeks") },
+            { value: "month", label: _t("Months") },
+            { value: "quarter", label: _t("Quarters") },
+            { value: "year", label: _t("Years") },
+        ];
+        const matching = this.getters.getOdooChartFieldMatching(chartId, filterId);
+        if (matching?.type === "datetime") {
+            all.unshift({ value: "hour", label: _t("Hours") });
+        }
+        const currentIndex = all.findIndex(
+            (granularity) => granularity.value === getMinGranularity(currentFilterValue)
+        );
+        return all.slice(Math.max(currentIndex - 2, 0), currentIndex + 1);
```

Comment:
"year to date" filter always have one year, but we can still select the "year" groupBy, which seems not relevant

Reply:
it can be relevant if you want to know the aggregated value since the beginning of the year